### DISCUSSION TITLE
meson: unable to find mpiifort

### DISCUSSION THREAD

**Discussion by dguibert** [NorESMhub/BLOM#206] [Q&A] [answered]
Got this when trying to configure with `intel-mpi@2019.7.217` and `intel@19.1.1.217` and `meson@0.55.3` (and newer 0.61.2)
````sh
mpiifort binary missing from cross or native file, or env var undefined.
Trying a default mpiifort fallback at mpiifort
mpiifort found: NO
Run-time dependency MPI for fortran found: NO (tried config-tool and system)

../meson.build:130:2: ERROR: Dependency "mpi" not found, tried config-tool and system
````
Do you know how to debug this?
@matsbn, @monsieuralok or others

a simple project as following fails:
````meson
project('mpitest', 'fortran')
dependency('mpi', language: 'fortran')
````

**Comment by monsieuralok** [comment]
@nordmoen please could you reply regarding meson build.

**Comment by nordmoen** [comment, marked as answer]
Hi, there seems to be [some peculiarities](https://github.com/mesonbuild/meson/issues/9637) in how Meson detects Intel MPI.

Use the following to get it to work:
```sh
$ FC=ifort I_MPI_F90=mpifort meson setup builddir
```
---
Just wanted to note as well, excellent issue @dguibert, it is super helpful with a small reproducible example!

**Comment by dguibert** [comment]
As you pointed out, Meson is broken detecting Intel MPI. As mentionned if `I_MPI_F90` is set to `mpiifort` itself, it enters in a infinite loop since the wrapper is calling himself.

So `I_MPI_F90` should be unset.

I really don't understand, the example works on Betzy, not on another cluster: same compiler and mpi version.

**Comment by nordmoen** [comment]
Don't know if you tried, but i set `I_MPI_F90` to `mpifort` and _not_ `mpiifort`

**Comment by dguibert** [comment]
On betzy, with module `iimpi/2020a`, `mpifort` is not found (Or I miss something).

I found the problem on the other cluster, the installation of Intel compilers defines custom options in `ifort.cfg`: `-Xlinker -rpath=...`, so `ifort -v` shows:
```
ifort version 19.1.1.217
ld    /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crt1.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crti.o /usr/lib/gcc/x86_64-redhat-linux/8/crtbegin.o --eh-frame-hdr --build-id -dynamic-linker /lib64/ld-linux-x86-64.so.2 -m elf_x86_64 -plugin /scratch/users/dguibert/s1-call3-blom/spack/opt/spack/linux-rhel8-x86_64/gcc-8.4.1/intel-19.1.1-q5oonts7ruca2ocddtunnb27v473xtdf/compilers_and_libraries_2020.1.217/linux/bin/intel64/../../lib/LLVMgold.so -o a.out [...]/compilers_and_libraries_2020.1.217/linux/compiler/lib/intel64 -Bdynamic -Bstatic -lifport -lifcoremt -limf -lsvml -Bdynamic -lm -Bstatic -lipgo -lirc -Bdynamic -lpthread -Bstatic -lsvml -Bdynamic -lc -lgcc -lgcc_s -Bstatic -lirc_s -Bdynamic -ldl -lc /usr/lib/gcc/x86_64-redhat-linux/8/crtend.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crtn.o
/scratch/users/dguibert/s1-call3-blom/spack/opt/spack/linux-rhel8-x86_64/gcc-8.4.1/intel-19.1.1-q5oonts7ruca2ocddtunnb27v473xtdf/compilers_and_libraries_2020.1.217/linux/compiler/lib/intel64_lin/for_main.o: In function `main':
for_main.c:(.text+0x2e): undefined reference to `MAIN__'
````
and exits with error code 1 which breaks Meson `find_external_program` in `find_config` from configtool class.

**Comment by nordmoen** [comment]
Sorry I was testing with `iompi`, indeed with `iimpi` it is `mpiifort`.

**Comment by dguibert** [comment]
Just to mention that a PR has been submitted to meson to solve this: https://github.com/mesonbuild/meson/pull/10056

**Reply by TomasTorsvik** [reply]
@dguibert 
Thanks for the update! Looks like the PR has been pending for a while ...
